internal/semaphore/semaphore.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package semaphore
+
+import (
+	"log"
+)
+
+type Semaphore interface {
+	//Acquire a slot in the semaphore
+	Acquire()
+
+	//Release a slot in the semaphore.
+	Release()
+
+	//GetLimit returns semaphore limit.
+	GetLimit() int
+
+	//GetCount returns current number of occupied slots in semaphore.
+	GetCount() int
+
+	//Done closes down the semaphore
+	Done()
+}
+
+type semaphore struct {
+	slots chan struct{}
+}
+
+func NewSemaphore(limit int) Semaphore {
+	if limit < 0 {
+		log.Panic("E! Semaphore size limit must not be negative")
+	}
+
+	sem := &semaphore{
+		slots: make(chan struct{}, limit),
+	}
+
+	for i := 0; i < cap(sem.slots); i++ {
+		sem.slots <- struct{}{}
+	}
+	
+	return sem
+}
+
+func (sem *semaphore) Acquire() {
+	<-sem.slots
+}
+
+// Release the acquired semaphore.
+func (sem *semaphore) Release() {
+	select {
+	case sem.slots <- struct{}{}:
+	default:
+		log.Panic("E! Semaphore released more than held")
```

Comment:
Why are we panicking here? I thought this was supposed to be for debugging. I do not want this change to break the agent just for debugging.